File: 75.颜色分类.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int i = -1, j = 0, k = nums.size();
        while(j < k){
            if(nums[j] == 0){
                ++i;
                swap(nums, i, j);
                ++j;
            } else if (nums[j] == 1)
                ++j;
            else{
                --k;
                swap(nums, j, k);
            }
        }
    }

    void swap(vector<int>& nums, int i, int j){
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
};

File: LeetCode 热题 HOT 100/75.颜色分类.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
// 快排思想
var sortColors = function (nums) {
    let i = 0,
        j = 0,
        k = nums.length - 1;
    while (j <= k) {
        if (nums[j] === 0) {
            [nums[i], nums[j]] = [nums[j], nums[i]];
            ++j;
            ++i;
        } else if (nums[j] === 1) {
            ++j;
        } else {
            [nums[j], nums[k]] = [nums[k], nums[j]];
            --k;
        }
    }
};